frontend/src/components/resourceMap/GraphView.tsx

```diff
@@ -0,0 +1,390 @@
+
+        // Only fit bounds when user hasn't moved viewport manually
+        if (!viewportMovedRef.current) {
+          flow.fitBounds(bounds, { padding: 0.2 });
+        }
+      });
+    });
+  }, [visibleGraph, aspectRatio, flow]);
+
+  // Reset after view change
+  useLayoutEffect(() => {
+    viewportMovedRef.current = false;
+  }, [selectedNodeId, groupBy, expandAll]);
+
+  const selectedNode = nodes.find((it: GraphNode) => it.id === selectedNodeId);
+
+  let selectedGroup: KubeGroupNode | undefined;
+  if (selectedNodeId) {
+    selectedGroup = findGroupContaining(visibleGraph, selectedNodeId) as any;
+  }
+  const highlights = useNodeHighlight(selectedNodeId);
+
+  const graphSize = getGraphSize(visibleGraph);
+  useEffect(() => {
+    if (expandAll && graphSize > 50) {
+      setExpandAll(false);
+    }
+  }, [graphSize]);
+
+  return (
+    <GraphViewContext.Provider
+      value={useMemo(
+        () => ({ nodeSelection: selectedNodeId, highlights, setNodeSelection: setSelectedNodeId }),
+        [selectedNodeId, setSelectedNodeId, highlights]
+      )}
+    >
+      <Box
+        sx={{
+          position: 'relative',
+          height: height ?? '800px',
+          display: 'flex',
+          flexDirection: 'row',
+          flex: 1,
+        }}
+      >
+        <CustomThemeProvider>
+          <Box
+            sx={() => ({
+              overflow: 'hidden',
+              display: 'flex',
+              flexDirection: 'column',
+              position: 'relative',
+              flexGrow: 1,
+              background: '#00000002',
+            })}
+          >
+            <Box padding={2} pb={0}>
+              <Box display="flex" gap={1} alignItems="center" mb={1} flexWrap="wrap">
+                <ResourceSearch
+                  resources={nodes.flatMap(it =>
+                    it.type === 'kubeObject' ? [it.data.resource] : []
+                  )}
+                  onSearch={resource => {
+                    setSelectedNodeId(resource.metadata.uid);
+                  }}
+                />
+
+                <NamespacesAutocomplete />
+
+                <GraphSourcesView
+                  sources={defaultSources}
+                  selectedSources={selectedSources}
+                  toggleSource={toggleSelection}
+                  sourceData={sourceData ?? new Map()}
+                />
+
+                {namespaces.size !== 1 && (
+                  <Chip
+                    label={t('Group By: {{ name }}', { name: t('Namespace') })}
+                    color={groupBy === 'namespace' ? 'primary' : undefined}
+                    variant={groupBy === 'namespace' ? 'filled' : 'outlined'}
+                    icon={groupBy === 'namespace' ? <Icon icon="mdi:check" /> : undefined}
+                    onClick={() => setGroupBy(it => (it === 'namespace' ? undefined : 'namespace'))}
+                    sx={{
+                      lineHeight: '1',
+                    }}
+                  />
```

Comment:
Should these be abstracted into their own components? Seems like we are repeating similar code.

Reply:
It looks like these were moved into a ChipToggleButton, resolving.